app/services/subscriptions/dates_service.rb

```diff
@@ -0,0 +1,304 @@
+# frozen_string_literal: true
+
+module Subscriptions
+  class DatesService < BaseService
+    def initialize(subscription, timestamp)
+      @subscription = subscription
+
+      # NOTE: timestamp is the billing day
+      #       It should be the end of the billing period + 1 day
+      #       When subscription is terminated, it is the termination day
+      @timestamp = timestamp
+
+      super(nil)
+    end
+
+    def from_date
+      return @from_date if @from_date
+
+      @from_date = case plan.interval.to_sym
+                   when :weekly
+                     weekly_from_date
+                   when :monthly
+                     monthly_from_date
+                   when :yearly
+                     yearly_from_date
+                   else
+                     raise NotImplementedError
+      end
+
+      # NOTE: On first billing period, subscription might start after the computed start of period
+      #       ei: if we bill on beginning of period, and user registered on the 15th, the invoice should
+      #       start on the 15th (subscription date) and not on the 1st
+      @from_date = subscription.started_at.to_date if @from_date < subscription.started_at
+
+      @from_date
+    end
+
+    def to_date
+      return @to_date if @to_date
+
+      @to_date = case plan.interval.to_sym
+                 when :weekly
+                   weekly_to_date
+                 when :monthly
+                   monthly_to_date
+                 when :yearly
+                   yearly_to_date
+                 else
+                   raise NotImplementedError
+      end
+
+      if subscription.terminated? && @to_date > subscription.terminated_at
+        # NOTE: When subscription is terminated, we cannot generate an invoice for a period after the termination
+        @to_date = if %i[pending active].include?(subscription.next_subscription&.status&.to_sym)
+          subscription.terminated_at.to_date - 1.day
```

Comment:
Not sure to understand `- 1 day` here.. 🤔 

For instance, if `next_subscription` is active for more than 1 day.

Reply:
It is to cover the upgrade case. When upgrading, we will have `previous_subscription.terminated_at = next_subscription.started_at` and we do not want to bill fees twice on both subscriptions.

That  said, you are right, we might have here if terminated_at == started_at on the terminated subscription. I'll fix this one.

*Note:* Keep in mind that fees for the period will be bill on the next plan
(That's something we need to challenge but it is the current behavior)